Parquet: Refactor decompression, including more complete datapage v2 support

wisp3rwind opened the pull request:
I went ahead on https://github.com/duckdb/duckdb/discussions/4437: This refactors parquet decompression to share code, with the additional benefit of supporting all compression algorithms on v2 data pages.
(The last commit is unrelated, but from my understanding this size should really be validated instead of trusting the snappy header. I could move that into a separate PR.)

The PR compiles, but I have not actually tested it up to now. Does this look like the right approach? Then, I would look into adding a few additional parquet files that test the newly supported combinations of compression algorithm/data page version.

This should be easiest to review commit-by-commit.

wisp3rwind committed:
parquet: identical signature for PreparePage and PreparePageV2

wisp3rwind committed:
Parquet: refactor ColumnReader::PreparePage

preparation to merge compression code with PreparePageV2

wisp3rwind committed:
Parquet: move decompression from PreparePage into separate function

wisp3rwind committed:
Parquet: in PreparePageV2, use the shared DecompressInternal

wisp3rwind committed:
Parquet: converge PreparePage and PreparePageV2

wisp3rwind committed:
Parquet: verify uncompressed size for snappy

can't trust the parquet page header here

wisp3rwind committed:
Parquet: code style

wisp3rwind commented:
> Thanks for the PR! Could you include some tests for the new functionality?

Sure! I just wanted to get some quick feedback (no thorough review) before spending the time on writing tests.

Mytherin reviewed (commented):
The code looks fine to me in principle but I would still really like to see tests added before merging.

wisp3rwind commented on extension/parquet/column_reader.cpp:
```diff
@@ -188,95 +188,93 @@ void ColumnReader::PrepareRead(parquet_filter_t &filter) {
+                                      idx_t dst_size) {
+	switch (codec) {
 	case CompressionCodec::UNCOMPRESSED:
+		D_ASSERT(false);
```
Done (not yet pushed).

wisp3rwind commented:
> The code looks fine to me in principle but I would still really like to see tests added before merging.

Absolutely; I've just not gotten around to finishing the tests (I have a WIP script to generate test data).

wisp3rwind commented:
Is it possible to only run a subset of the sqllogictests when testing locally? https://duckdb.org/dev/testing does not mention anything.

Mytherin commented:
> For example, if we wanted to execute only the above test, we would run the command unittest test/sql/projection/test_simple_projection.test. If we wanted to run all tests in a specific directory, we would run the command unittest "[projection]".

wisp3rwind committed:
parquet: address review (D_ASSERT -> InternalException)

wisp3rwind committed:
parquet: add data and test for various compression algorithms

Generate a relatively complex table, since v2 data pages compress
repetition/definition levels differently (i.e. not at all) so we should
be sure to have columns with and without NULLs/nesting/variable-length
fields.

wisp3rwind committed:
parquet: in SQL tests, unskip unsupported formats and add statement error mark

should slightly increase coverage by hitting the error handler for
unsupported formats

Mytherin commented:
Thanks! LGTM
